subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy

```diff
@@ -815,6 +815,128 @@ class JavaCompileIntegrationTest extends AbstractIntegrationSpec {
         !file("build/classes/java/main/ignored/io/ignored/IgnoredExample.class").exists()
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/20394")
+    @Requires(TestPrecondition.JDK16_OR_LATER)
+    def "class snapshotting ignores changes to a record implementation"() {
+        given:
+        settingsFile << "include 'a', 'b'"
+
+
+        file('a/build.gradle') << """
+            apply plugin: 'java'
+        """
+        file('a/src/main/java/Foo.java') << 'record Foo(String property) {}'
+
+        file('b/build.gradle') << '''
+            apply plugin: 'java'
+
+            dependencies {
+                implementation project(':a')
+            }
+        '''
+        file('b/src/main/java/Bar.java') << 'class Bar { public void useFoo(Foo foo) { } }'
+
+        when:
+        // Run with --debug to ensure that class snapshotting didn't fail.
+        succeeds ":b:compileJava", "--debug"
+
+        then:
+        executedAndNotSkipped(":b:compileJava")
+
+        when:
+        // Change internal implementation, but not API.
+        file('a/src/main/java/Foo.java').delete()
+        file('a/src/main/java/Foo.java') << '''
+record Foo(String property) {
+    public String property() {
+        return property + "!";
+    }
+}
+'''
+
+        succeeds ":b:compileJava"
+
+        then:
+        skipped(":b:compileJava")
+
+        when:
+        // Add new property, breaks API.
+        file('a/src/main/java/Foo.java').delete()
+        file('a/src/main/java/Foo.java') << '''
+record Foo(String property, int newProperty) {
+    public String property() {
+        return property + "!";
+    }
+}
+'''
+
+        succeeds ":b:compileJava"
+
+        then:
+        executedAndNotSkipped(":b:compileJava")
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/20394")
+    @Requires(TestPrecondition.JDK17_OR_LATER)
+    def "class snapshotting changes when a sealed modifier is changed"() {
```

Comment:
Would it make more sense to put these two tests in `AbstractJavaCompileAvoidanceIntegrationSpec`?  (And if so, we should change the test descriptions to match the style in those classes - something like "recompiles when a sealed modifier is changed" or some such)

Reply:
That's a much better spot, thanks. I've done this.